pkg/discovery/backends.go

```diff
@@ -39,6 +40,65 @@ func parse(rawurl string) (string, string) {
 	return parts[0], parts[1]
 }
 
+// ParseAdvertise parses the --cluster-advertise daemon config which accepts
+// <ip-address>:<port> or <interface-name>:<port>
+func ParseAdvertise(store, advertise string) (string, error) {
+	var (
+		iface *net.Interface
+		addrs []net.Addr
+		err   error
+	)
+
+	addrPort := strings.Split(advertise, ":")
+
+	if len(addrPort) > 2 {
+		return "", fmt.Errorf("cluster advertise config (%s) doesnt match <ipv4-address>:<port> format", advertise)
+	}
+
+	addrStr := addrPort[0]
+	ip := net.ParseIP(addrStr)
+
+	// If advertise is a valid ip-address, use it as is
+	if ip != nil {
+		addrStr = ip.String()
+		goto success
+	}
+
+	// If advertise is a valid interface name, get the valid ipv4 address and use it to advertise
+	iface, err = net.InterfaceByName(addrStr)
+	if err != nil {
+		return "", fmt.Errorf("invalid cluster advertise ip-address or interface name (%s) : %v", advertise, err)
+	}
+
+	addrs, err = iface.Addrs()
+	if err != nil {
+		return "", fmt.Errorf("error deriving advertise ip-address in interface (%s) : %v", advertise, err)
+	}
+
+	if addrs == nil || len(addrs) == 0 {
+		return "", fmt.Errorf("no available advertise ip-address in interface (%s)", advertise)
+	}
+
+	for _, a := range addrs {
+		ip, _, err := net.ParseCIDR(a.String())
+		if err != nil {
+			return "", fmt.Errorf("error deriving advertise ip-address in interface (%s) : %v", advertise, err)
+		}
+		if ip.To4() == nil || ip.IsLoopback() {
+			continue
+		}
+		addrStr = ip.String()
+		goto success
```

Comment:
Never said all `goto`s were bad but they should be used cautiously and only when necessary (retry mechanism, etc.). Here it is totally unnecessary and the fact that it's still easily readable does not mean that the `goto` does not decrease the overall readability. Anyway, I made my case for the use of `goto`s :wink: